src/v/security/gssapi_authenticator.cc

```diff
@@ -0,0 +1,514 @@
+    OM_uint32 minor_status{};
+    OM_uint32 major_status{};
+
+    gss::buffer_view service_name{principal};
+    gss::name server_name{};
+
+    major_status = ::gss_import_name(
+      &minor_status, &service_name, GSS_C_NT_HOSTBASED_SERVICE, &server_name);
+
+    vlog(
+      seclog.trace,
+      "gss {} import name {}:{} - {}",
+      _state,
+      major_status,
+      minor_status,
+      principal);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("import name", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+
+    gss_key_value_element_desc elem{.key = "keytab", .value = keytab.c_str()};
+    gss_key_value_set_desc store{.count = 1, .elements = &elem};
+
+    major_status = ::gss_acquire_cred_from(
+      &minor_status,
+      server_name,
+      0,
+      GSS_C_NO_OID_SET,
+      GSS_C_ACCEPT,
+      &store,
+      &_server_creds,
+      NULL,
+      NULL);
+
+    vlog(
+      seclog.info,
+      "gss {} acquire cred {}:{} - {}",
+      _state,
+      major_status,
+      minor_status,
+      keytab);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("gss acquire cred", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+
+    gss::buffer token;
+    major_status = gss_export_cred(&minor_status, _server_creds, &token);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("gss acquire cred", major_status, minor_status);
+        _state = state::failed;
+        return errc::invalid_scram_state;
+    }
+    vlog(
+      seclog.trace,
+      "gss {} gss_export_cred: {}",
+      _state,
+      hex(bytes_view{token}));
+
+    gss::buffer display_name;
+    major_status = ::gss_display_name(
+      &minor_status, server_name, &display_name, &_doid);
+    if (major_status != GSS_S_COMPLETE) {
+        display_status("gss server name", major_status, minor_status);
+    } else {
+        vlog(
+          seclog.trace,
+          "gss {} server name {}:{} = {}",
+          _state,
+          major_status,
+          minor_status,
+          std::string_view{display_name});
+    }
+    _state = state::more;
+    return outcome::success();
+}
+
+result<bytes> gssapi_authenticator::more(bytes_view auth_bytes) {
+    gss::buffer_view recv_tok{auth_bytes};
+    gss::buffer send_tok;
+
+    vlog(
+      seclog.trace,
+      "recv_tok: length: {} - {}",
+      auth_bytes.length(),
+      hex{auth_bytes});
+
+    if (auth_bytes.length() == 0) {
+        if (auto res = finish(); res.has_error()) {
+            return res.assume_error();
+        }
+        return bytes{};
+    }
+
+    OM_uint32 major_status{}, minor_status{}, ret_flags{};
+
+    major_status = ::gss_accept_sec_context(
+      &minor_status,
+      &_context,
+      _server_creds,
+      &recv_tok,
+      GSS_C_NO_CHANNEL_BINDINGS,
+      &_client_name,
+      &_doid,
```

Comment:
Is this correct?  I think this is for what security mechanism was used and this was populated by `gss_display_name` which is the type of the returned name.

Reply:
I'm referring to passing `_doid` to `gss_accept_sec_context`

Reply:
I've changed this.